find-simdoc-cli/src/minhash_acc.rs

```diff
@@ -109,36 +133,84 @@ fn main() -> Result<(), Box<dyn Error>> {
         sketches
     };
 
-    let jac_dists = {
+    let tmp_path = {
+        let mut tmp_path = tmp_dir.unwrap_or_else(env::temp_dir);
+        tmp_path.push("tmp.jac_dist");
+        tmp_path
+    };
+
+    let possible_pairs = {
+        let start = Instant::now();
+
         let possible_pairs = features.len() * (features.len() - 1) / 2;
         eprintln!("Computing exact Jaccard distances for {possible_pairs} pairs...");
-        let start = Instant::now();
-        let mut jac_dists = Vec::with_capacity(possible_pairs);
-        for i in 0..features.len() {
-            if (i + 1) % 100 == 0 {
-                eprintln!("Processed {}/{}...", i + 1, features.len());
+
+        let tmp_file_size = possible_pairs * mem::size_of::<f64>();
+        let offsets = {
+            let mut offset = 0;
+            let mut offsets = Vec::with_capacity(features.len());
+            for i in 0..features.len() {
+                offsets.push(offset);
+                offset += features.len() - i - 1;
             }
-            let x = &features[i];
-            for y in features.iter().skip(i + 1) {
-                jac_dists.push(find_simdoc::lsh::jaccard_distance(
-                    x.iter().clone(),
-                    y.iter().clone(),
-                ));
+            assert_eq!(offset, possible_pairs);
+            offsets
+        };
+
+        {
+            let processed = Mutex::new(0usize);
+            let writer = Mutex::new(File::create(&tmp_path)?);
+
+            // Creates a file object of size tmp_file_size bytes.
+            {
+                let mut w = writer.lock().unwrap();
+                w.write_at(tmp_file_size as u64 - 1, &[0])?;
             }
+
+            eprintln!(
+                "Created a tmp file of {} GiB, at {:?}",
+                tmp_file_size as f64 / (1024. * 1024. * 1024.),
+                &tmp_path
+            );
+
+            (0..features.len()).into_par_iter().for_each(|i| {
```

Comment:
Is it possible to parallelize possible feature pairs (x, y) instead of parallelizing single features?
If possible, the runtime would become faster since the sizes which each thread processes are much different, and the code would be more clear.

Reply:
I think It is not be efficient because:

Computed distances are written in a file with a chunk of `jac_dists`.
Since `positioned_io::WriteAt` is not implemented for `BufWriter`, such a manual buffering is essential, and to implement this, looping for only `x` is convenient.

Reply:
btw, Rayon will automatically assign task such that the performance is maximized.

https://github.com/rayon-rs/rayon#parallel-iterators-and-more

>  Parallel iterators take care of deciding how to divide your data into tasks; it will dynamically adapt for maximum performance.

I believe Rayon.

Reply:
I agreed. I tested my idea, and the runtime was 10 times slower.
Thank you for your explanation.